### PR TITLE
Switch from API keys to oauth

### PR DESCRIPTION
**Pull request by Pierlou**

### CODE REVIEW

**Review by JM-D — COMMENTED** [review]
Cela me parait bon pour le traitement du token OAuth2

File: utils.py [hunk 1]
```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
 LOG_PATH = "./logs/"
 
 
+class Meteo_client(object):
```
Comment (HAEKADI) [inline utils.py]:
```suggestion
class MeteoClient(object):
```

For PEP8 ;)

File: utils.py [hunk 2]
```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
...
+            self.session.headers.update({'Authorization': f'Bearer {token}'})
+            logging.info("Fetched and saved new token")
+        else:
+            logging.info("Checking token.json")
+            with open("token.json", "r") as f:
+                token = json.load(f)["token"]
+            self.session.headers.update({'Authorization': f'Bearer {token}'})
+            response = self.session.request(
+                "GET",
+                # this is just an example URL to check potential expiration
```
Comment (HAEKADI) [inline utils.py]: What does this mean ?

File: utils.py [hunk 3]
```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
...
+                "https://public-api.meteofrance.fr/previnum/DPPaquetWAVESMODELS/models/MFWAM/grids",
```
Comment (HAEKADI) [inline utils.py]: Same for this endpoint. I would define it for the class.

File: utils.py [hunk 4]
```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
...
+            )
+            if self.token_has_expired(response):
+                logging.warning("Token has expired, fetching a fresh one")
+                os.remove("token.json")
+                self.obtain_token()
+
+
+meteo_client = Meteo_client()
```
Comment (HAEKADI) [inline utils.py]:
```suggestion
meteo_client = MeteoClient()
```

File: utils.py [hunk 5]
```diff
@@ -142,12 +209,12 @@ def download_url(url, meta_urls, retry, current_folder) -> None:
...
-            headers = meta_urls[url + ":headers"]
-            with requests.get(
+            with meteo_client.get(
                 url,
                 stream=True,
-                headers=headers,
-                timeout=60
+                timeout=60,
+                # do we actually need these headers? response content type is binary
+                headers={"Content-Type": "application/json; charset=utf-8"},
```
Comment (HAEKADI) [inline utils.py:219]: If they're universal, you can just integrate the headers in the class method `get` maybe.

File: utils.py [hunk 6]
```diff
@@ -234,9 +301,9 @@ def remove_and_create_folder(folder_path: str, toCreate: bool) -> None:
         os.makedirs(folder_path)
 
 
-def check_if_data_available(batches: list, url: str, apikey: str) -> list:
+def check_if_data_available(batches: list, url: str) -> list:
```
Comment (HAEKADI) [inline utils.py]: I would expect the function to return a bool given it's name.
i think something like `get_new_data` or `get_new_batches`  would make more sense. What do you think ?

File: utils.py (hunk 5 above)
Comment (Pierlou) [reply utils.py:219]: They're not, these headers are used when getting the files, but the same client is used to ping the API to check for the availability

File: utils.py [hunk 7]
```diff
@@ -47,6 +49,71 @@ class File(TypedDict):
...
+                return True
+        return False
+
+    def obtain_token(self):
+        # if threads are synchronous (for example on start), this should space them out
+        time.sleep(random.randint(1, 5))
+        if "token.json" not in os.listdir():
+            # Obtain new token
+            access_token_response = requests.post(
+                "https://portail-api.meteofrance.fr/token",
```
Comment (HAEKADI) [inline utils.py]: I would define this endpoint url for as a class param. It would be easier to change and keep track of.

File: utils.py (hunk 2 above)
Comment (Pierlou) [reply utils.py]: It could be any valid URL from the same API, this call is only here to check if the current token is still valid

**Review by HAEKADI — APPROVED** [review]

**Pull request merged**